### PR TITLE
Implement the base `FetcherService`

### PR DESCRIPTION
**Pull request by jamietanna**

### CODE REVIEW

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java [hunk 1]
```diff
@@ -1,6 +1,6 @@
 package uk.gov.api.springboot.infrastructure;
 
-import static com.github.valfirst.slf4jtest.Assertions.*;
+import static com.github.valfirst.slf4jtest.Assertions.assertThat;
```
Comment (jamietanna) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/infrastructure/ScheduleHandlerTest.java:3]: should be undone

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/domain/services/fetcher/FetcherService.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      try {
+        var fetched = fetcher.fetch(entry.baseUrl());
+        LOGGER.info("Retrieved {} APIs from the {} API", fetched.size(), entry.id());
+        return fetched;
+      } catch (Fetcher.VersionNotSupportedException e) {
+        // ignore
+      } catch (Fetcher.ClientErrorException e) {
+        if (LOGGER.isErrorEnabled()) {
+          LOGGER.error(
+              "A client error caused retrieve of APIs from %s to fail".formatted(entry.id()), e);
```
Comment (nahteb) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/domain/services/fetcher/FetcherService.java:42]:
```suggestion
              "A client error caused retrieval of APIs from %s to fail".formatted(entry.id()), e);
```

File: examples/java/spring-boot/src/main/java/uk/gov/api/springboot/domain/services/fetcher/FetcherService.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        }
+        throw e;
+      } catch (Fetcher.TemporaryErrorException e) {
+        if (LOGGER.isErrorEnabled()) {
+          LOGGER.error(
+              "A temporary error caused retrieve of APIs from %s to fail".formatted(entry.id()), e);
```
Comment (nahteb) [inline examples/java/spring-boot/src/main/java/uk/gov/api/springboot/domain/services/fetcher/FetcherService.java:48]:
```suggestion
              "A temporary error caused retrieval of APIs from %s to fail".formatted(entry.id()), e);
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/domain/services/FetcherServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+        when(v1Fetcher.fetch(any())).thenThrow(exception);
+
+        assertThatThrownBy(() -> fetcherService.fetch(entry))
+            .isInstanceOf(Fetcher.ClientErrorException.class);
+
+        assertThat(logger)
+            .hasLogged(
+                LoggingEvent.error(
+                    exception,
+                    "A client error caused retrieve of APIs from %s to fail"
```
Comment (nahteb) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/domain/services/FetcherServiceTest.java:226]:
```suggestion
                    "A client error caused retrieval of APIs from %s to fail"
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/domain/services/FetcherServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+        when(v1Fetcher.fetch(any())).thenThrow(exception);
+
+        assertThatThrownBy(() -> fetcherService.fetch(entry))
+            .isInstanceOf(Fetcher.TemporaryErrorException.class);
+
+        assertThat(logger)
+            .hasLogged(
+                LoggingEvent.error(
+                    exception,
+                    "A temporary error caused retrieve of APIs from %s to fail"
```
Comment (nahteb) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/domain/services/FetcherServiceTest.java:257]:
```suggestion
                    "A temporary error caused retrieval of APIs from %s to fail"
```

File: examples/java/spring-boot/src/test/java/uk/gov/api/springboot/domain/services/FetcherServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+          throws Fetcher.VersionNotSupportedException, Fetcher.ClientErrorException,
+              Fetcher.TemporaryErrorException {
+        when(v1Fetcher.fetch(any())).thenThrow(new Fetcher.VersionNotSupportedException());
+        when(v2Fetcher.fetch(any())).thenThrow(new Fetcher.VersionNotSupportedException());
+        when(v3Fetcher.fetch(any())).thenThrow(new Fetcher.VersionNotSupportedException());
+
+        assertThatThrownBy(() -> fetcherService.fetch(entry))
+            .isInstanceOf(Fetcher.ClientErrorException.class);
+
+        System.out.println(logger.getLoggingEvents());
```
Comment (nahteb) [inline examples/java/spring-boot/src/test/java/uk/gov/api/springboot/domain/services/FetcherServiceTest.java:191]: is this needed?

**Review by nahteb — COMMENTED** [review]
just a couple of small things, otherwise LGTM

**Review by nahteb — APPROVED** [review]

**Pull request merged**